---
title: "`r gsub('.Rmd', ' ', gsub('_', ' ', knitr::current_input()))`"
author: "`r Sys.getenv('USER')`"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    toc: yes
---
  


```{r setup, include = FALSE}

library(tidyverse)
library(scales)
library(kableExtra)

```


```{r}

transcript_coord_input_file <- "example_transcript_coordinates.tsv"

gene_model_gtf_file <- "NHA_hTERT_DRNA_20220609_flair_collapse_output.isoforms.gtf.gz"
```


# Import modification results
```{r}

transcript_coord_input <- read_tsv(transcript_coord_input_file) %>%
  rename(transcript_position = position)

```

# Import transcript-genome info
```{r}
gtf_column_names <- c("seqname", "source", "feature", "start", "end", 
                        "score", "strand", "frame", "attributes")

gtf_raw <- read_tsv(gene_model_gtf_file,
                     col_names = gtf_column_names)

chr_in_order <- paste0("chr", c(1:22, "X", "M"))

gtf <-  gtf_raw %>%
  filter(feature == "exon") %>%
  rename(open_end = end) %>%
  mutate(end = open_end +1) %>% # compensate for coords being end-inclusive
  mutate(gene_id = str_remove(attributes, '";.*$') %>% str_remove('^.*"'),
         transcript_id = str_remove(attributes, '"; exon.*$') %>% str_remove('^.*transcript_id "'),
         exon_number = str_remove(attributes, '";$') %>% str_remove('^.*exon_number "') %>% as.numeric(),
         seqname = factor(seqname, levels = chr_in_order),
         exon_length = abs(end-start)
         )

gtf_relevant <- gtf %>%
  filter(transcript_id %in% transcript_coord_input$transcript_id) %>%
  select(seqname, start, end, gene_id, transcript_id, exon_number, exon_length, strand)


```


# Cacluate genome position of transcripts
```{r}

gtf_relevant_with_pos <- gtf_relevant %>%
  arrange(seqname, transcript_id, exon_number) %>%
  group_by(transcript_id) %>%
  mutate(exon_start_pos_in_transcript = cumsum(exon_length)-exon_length+1,
         exon_end_pos_in_transcript = cumsum(exon_length)) %>%
  rowwise() %>%
  mutate(genome_position = list(start:(end-1)),
         transcript_position = list(exon_start_pos_in_transcript:exon_end_pos_in_transcript)) %>%
  select(seqname, gene_id, transcript_id, exon_number, genome_position, transcript_position, strand) 

gtf_relevant_with_pos_list <- gtf_relevant_with_pos %>%
  group_by(seqname, transcript_id, strand) %>%
  summarize(genome_pos = case_when(
    strand == "+" ~ unlist(genome_position)  %>% list,
    strand == "-" ~ unlist(genome_position) %>% rev() %>% list),
    transcript_pos = unlist(transcript_position) %>% list) %>% 
  distinct()

```

# convert coordinates
```{r}

conversion_results_raw <- left_join(transcript_coord_input, 
          gtf_relevant_with_pos_list,
           by = "transcript_id") %>%
    rowwise() %>%
  mutate(genome_position_of_modification = genome_pos[transcript_position])

conversion_results <- conversion_results_raw %>%
  select(gene_id, transcript_id, transcript_position,
         chr = seqname, genome_position = genome_position_of_modification, strand) 
   

time_stamp_for_file_name <- format(Sys.time(), "%Y%m%d_%H%M%S")

write_tsv(conversion_results, paste0("genome_coords.", time_stamp_for_file_name, ".tsv"))

```


